skipchain/skipchain.go

```diff
@@ -144,331 +173,427 @@ func (s *Service) ProposeSkipBlock(psbd *ProposeSkipBlock) (network.Message, one
+	if err := s.SendRaw(node,
+		&GetBlock{unknown}); err != nil {
+		return nil, errors.New("Couldn't get updated block: " + known.Short())
+	}
+	var block *SkipBlock
+	select {
+	case block = <-s.blockRequests[string(unknown)]:
+		log.Lvl3("Got block", block)
+		delete(s.blockRequests, string(unknown))
+	case <-time.After(time.Millisecond * time.Duration(propagateTimeout)):
+		delete(s.blockRequests, string(unknown))
+		return nil, errors.New("Couldn't get updated block in time: " + unknown.Short())
+	}
+	return block, nil
+}
+
+// forwardSignature receives a signature request of a newly accepted block.
+// It only needs the 2nd-newest block and the forward-link.
+func (s *Service) forwardSignature(env *network.Envelope) {
+	err := func() error {
+		fs, ok := env.Msg.(*ForwardSignature)
+		if !ok {
+			return errors.New("Didn't receive a ForwardSignature")
+		}
+		if fs.TargetHeight >= len(fs.Newest.BackLinkIDs) {
+			return errors.New("This backlink-height doesn't exist")
+		}
+		target := s.Sbm.GetByID(fs.Newest.BackLinkIDs[fs.TargetHeight])
+		if target == nil {
+			return errors.New("Didn't find target-block")
+		}
+		data, err := network.Marshal(fs)
+		if err != nil {
+			return err
+		}
+		// TODO: is this really signed by target.roster?
+		sig, err := s.startBFT(bftFollowBlock, target.Roster, fs.ForwardLink.Hash, data)
+		if err != nil {
+			return errors.New("Couldn't get signature")
+		}
+		target.AddForward(&BlockLink{fs.ForwardLink.Hash, sig.Sig})
+		s.startPropagation([]*SkipBlock{target})
+		return nil
+	}()
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (s *Service) getBlock(env *network.Envelope) {
+	gb, ok := env.Msg.(*GetBlock)
 	if !ok {
-		log.Error("Couldn't convert to SkipBlock")
+		log.Error("Didn't receive GetBlock")
+		return
+	}
+	sb := s.Sbm.GetByID(gb.ID)
+	if sb == nil {
+		log.Error("Did not find block")
 		return
 	}
-	if err := sb.VerifySignatures(); err != nil {
+	if i, _ := sb.Roster.Search(s.ServerIdentity().ID); i < 0 {
+		log.Lvl3("Not responsible for that block, recursing")
+		var err error
+		sb, err = s.getUpdateBlock(sb, sb.Hash)
+		if err != nil {
+			log.Error(err)
+			sb = nil
+		}
+	}
+	if err := s.SendRaw(env.ServerIdentity, &GetBlockReply{sb}); err != nil {
 		log.Error(err)
+	}
+}
+
+func (s *Service) getBlockReply(env *network.Envelope) {
+	gbr, ok := env.Msg.(*GetBlockReply)
+	if !ok {
+		log.Error("Didn't receive GetBlock")
 		return
 	}
-	s.storeSkipBlock(sb)
-	log.Lvlf3("Stored skip block %+v in %x", *sb, s.Context.ServerIdentity().ID[0:8])
+	if err := s.Sbm.VerifyLinks(gbr.SkipBlock); err != nil {
+		log.Error("Received invalid skipblock: " + err.Error())
+	}
+	id := s.Sbm.Store(gbr.SkipBlock)
+	s.save()
+	log.Lvl3("Sending block to channel")
+	s.blockRequests[string(id)] <- gbr.SkipBlock
+}
+
+// verifyFollowBlock makes sure that a signature-request for a forward-link
+// is valid.
+func (s *Service) bftVerifyFollowBlock(msg []byte, data []byte) bool {
+	err := func() error {
```

Comment:
oh, sorry. The advantage here is that you can have only one `log.Error`, instead of having to `log.Error(); return` for every if-clause that you write.